src/Client/ClientBase.cpp

```diff
@@ -0,0 +1,1367 @@
+            /// Client-side exception during query execution can result in the loss of
+            /// sync in the connection protocol.
+            /// So we reconnect and allow to enter the next query.
+            if (!connection->checkConnected())
+                connect();
+        }
+    }
+    while (true);
+
+    if (isNewYearMode())
+        std::cout << "Happy new year." << std::endl;
+    else if (isChineseNewYearMode(local_tz))
+        std::cout << "Happy Chinese new year. 春节快乐!" << std::endl;
+    else
+        std::cout << "Bye." << std::endl;
+}
+
+
+void ClientBase::runNonInteractive()
+{
+    if (!queries_files.empty())
+    {
+        auto process_multi_query_from_file = [&](const String & file)
+        {
+            auto text = getQueryTextPrefix();
+            String queries_from_file;
+
+            ReadBufferFromFile in(file);
+            readStringUntilEOF(queries_from_file, in);
+
+            text += queries_from_file;
+            return processMultiQuery(text);
+        };
+
+        /// Read all queries into `text`.
+        for (const auto & queries_file : queries_files)
+        {
+            for (const auto & interleave_file : interleave_queries_files)
+                if (!process_multi_query_from_file(interleave_file))
+                    return;
+
+            if (!process_multi_query_from_file(queries_file))
+                return;
+        }
+
+        return;
+    }
+
+    String text;
+    if (is_multiquery)
+        text = getQueryTextPrefix();
+
+    if (config().has("query"))
+    {
+        text += config().getRawString("query"); /// Poco configuration should not process substitutions in form of ${...} inside query.
+    }
+    else
+    {
+        /// If 'query' parameter is not set, read a query from stdin.
+        /// The query is read entirely into memory (streaming is disabled).
+        ReadBufferFromFileDescriptor in(STDIN_FILENO);
+        readStringUntilEOF(text, in);
+    }
+
+    if (query_fuzzer_runs)
+        processWithFuzzing(text);
+    else
+        processQueryText(text);
+}
+
+
+static void clearTerminal()
+{
+    /// Clear from cursor until end of screen.
+    /// It is needed if garbage is left in terminal.
+    /// Show cursor. It can be left hidden by invocation of previous programs.
+    /// A test for this feature: perl -e 'print "x"x100000'; echo -ne '\033[0;0H\033[?25l'; clickhouse-client
+    std::cout << "\033[0J"
+                    "\033[?25h";
+}
+
+
+static void showClientVersion()
+{
+    std::cout << DBMS_NAME << " client version " << VERSION_STRING << VERSION_OFFICIAL << "." << std::endl;
+}
+
+
+int ClientBase::main(const std::vector<std::string> & /*args*/)
+{
+    UseSSL use_ssl;
+
+    std::cout << std::fixed << std::setprecision(3);
+    std::cerr << std::fixed << std::setprecision(3);
+
+    if (is_interactive)
+    {
+        clearTerminal();
+        showClientVersion();
+    }
+
+    return mainImpl();
+}
```

Comment:
> because LocalServer has it's own main

No, it doesn't. `main()` is only in ClientBase.

> so clickhouse-local does non print its version on start

No, there is a different reason -- because `is_interactive` is initialized too late. I'll fix.